fix: updated navigation event handling

jrandolf-google opened the pull request:
This PR separates navigation handling and lifecycle handling in Puppeteer:
- The new `NavigationWatcher` strictly handles navigation events and has been simplified drastically to only watch for the navigation response which appears after the request (hence the navigation) has finished.
- There is also some refactoring for tidiness.

OrKoN commented on src/common/Page.ts:
```diff
@@ -1871,7 +1871,9 @@ export class Page extends EventEmitter {
       this.waitForNavigation(options),
-      this.#client.send('Page.reload'),
+      this.evaluate(() => {
+        window.location.reload();
```
window.location might be overloaded by the page? Why not keep using Page.reload?

OrKoN replied:
another thing discovered: `window.location.reload` would not work for data URLs.

OrKoN commented on src/common/DOMWorld.ts:
```diff
@@ -284,17 +284,10 @@ export class DOMWorld {
-      watcher.timeoutOrTerminationPromise(),
-      watcher.lifecyclePromise(),
-    ]);
-    watcher.dispose();
```
do we still need dispose?

jrandolf replied:
No, all watchers dispose themselves on error or return.

jrandolf committed:
fix: updated navigation event handling

OrKoN committed:
feat: use CDP's auto-attach mechanism (#8520)

* feat: use CDP's auto-attach mechanism

In this PR, we refactor Puppeteer to make use of the CDP's auto-attach mechanism. This allows the backend to pause
new targets and give Puppeteer a chance to configure them properly. This fixes the flakiness related to dealing with
OOPIFs and should fix some other issues related to the network interception and navigations. If those are not fixed completely by this PR, the PR serves a solid base for fixing them.

Closes https://github.com/puppeteer/puppeteer/issues/8507, https://github.com/puppeteer/puppeteer/issues/7990
Unlocks https://github.com/puppeteer/puppeteer/issues/3667

BREAKING CHANGE: With Chromium, Puppeteer will now attach to page/iframe targets immediately. Browser.connect requires an explicit product name when connecting to Firefox since Firefox does not support CDP's auto-attach.

OrKoN committed:
fix: address flakiness in frame handling (#8688)

When we attach to a frame, we send a call to get
the page frame tree from CDP. Based on the tree data
we look up the parent frame if parentId is provided.
The problem is that the call to get the page frame
tree could take arbitrary time and the calls for the
parent and child frames might happen at the same time.
So the situation where the frame tree for the child frame
is resolved before the parent frame is known is fairly
common.

This PR addresses the issue by awaiting for the parent
frame id before attempting to register a child frame.

OrKoN committed:
feat: add Dockerfile (#8315)

This PR adds an official Dockerfile for Puppeteer. The content of the Dockerfile is practically the same as documented in troubleshooting.md:

1) It installs chrome-stable and dependencies via apt-get.
2) it installs a local Puppeteer build into the docker user's home folder.
3) configures required permissions for the user.
4) outputs licenses into the THIRD_PARTY_NOTICES file.

The local Puppeteer build is created by `docker/pack.sh` which is meant to be used in CI. This PR also includes a GitHub action that would build a docker image and run a smote test inside of it.

The next step would be actually publishing the docker image from GitHub Actions to GitHub Registry.

OrKoN committed:
chore: fix command name in tot-ci.yml (#8698)

dependabot[bot] committed:
chore(deps): bump ws from 8.8.0 to 8.8.1 (#8672)

Bumps [ws](https://github.com/websockets/ws) from 8.8.0 to 8.8.1.
- [Release notes](https://github.com/websockets/ws/releases)
- [Commits](https://github.com/websockets/ws/compare/8.8.0...8.8.1)

---
updated-dependencies:
- dependency-name: ws
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

OrKoN committed:
docs: fix a typo in waitForSelector (#8701)

dependabot[bot] committed:
chore(deps): bump actions/setup-node from 3.3.0 to 3.4.1 (#8671)

Bumps [actions/setup-node](https://github.com/actions/setup-node) from 3.3.0 to 3.4.1.
- [Release notes](https://github.com/actions/setup-node/releases)
- [Commits](https://github.com/actions/setup-node/compare/v3.3.0...v3.4.1)

---
updated-dependencies:
- dependency-name: actions/setup-node
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

OrKoN committed:
fix: disable AcceptCHFrame (#8706)

Closes: #8479

OrKoN committed:
chore: refactor init to make sure it works with paused targets (#8710)

dependabot[bot] committed:
chore(deps): bump github/codeql-action from 2.1.12 to 2.1.17 (#8709)

Bumps [github/codeql-action](https://github.com/github/codeql-action) from 2.1.12 to 2.1.17.
- [Release notes](https://github.com/github/codeql-action/releases)
- [Changelog](https://github.com/github/codeql-action/blob/main/CHANGELOG.md)
- [Commits](https://github.com/github/codeql-action/compare/27ea8f8fe5977c00f5b37e076ab846c5bd783b96...0c670bbf0414f39666df6ce8e718ec5662c21e03)

---
updated-dependencies:
- dependency-name: github/codeql-action
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

OrKoN committed:
fix: use loaderId to reduce test flakiness (#8717)

OrKoN committed:
chore: publish a Docker image (#8702)

This PR adds publishing of a Docker image to ghcr.io.

OrKoN commented on src/common/FrameManager.ts:
```diff
@@ -206,56 +207,16 @@ export class FrameManager extends EventEmitter {
-        throw error;
-      }
-    }
+    await this.#navigate(url, referer, frame._id);
```
I think we should call waitForFrameNavigation before we call navigate?

OrKoN commented on test/src/navigation.spec.ts:
```diff
@@ -116,11 +116,19 @@ describe('navigation', function () {
           false
         );
       });
+      // TODO: There is a determinism problem here. Because the history API is
```
is this comment still relevant?

OrKoN committed:
feat: detect Firefox in connect() automatically (#8718)

This PR implements automatic detection of the Firefox product when the `.connect()` method is used. This partially undoes the breaking change in https://github.com/puppeteer/puppeteer/pull/8520 but it's also a breaking change on its own since we don't accept an explicit product name anymore (it does not look like it was used anyway).
